contrib/api_incubator/src/main/java/io/opentelemetry/trace/SpanEx.java

```diff
@@ -0,0 +1,201 @@
+    return span;
+  }
+
+  public void setStatus(Status status) {
+    span.setStatus(status);
+  }
+
+  public void updateName(String name) {
+    span.updateName(name);
+  }
+
+  public SpanContext getContext() {
+    return span.getContext();
+  }
+
+  public boolean isRecording() {
+    return span.isRecording();
+  }
+
+  public void end() {
+    span.end();
+  }
+
+  public void end(EndSpanOptions endOptions) {
+    span.end(endOptions);
+  }
+
+  public static Builder newBuilder(Tracer tracer, String spanName) {
+    return new Builder(tracer, spanName);
+  }
+
+  public static class Builder {
+
+    private final Span.Builder builder;
+
+    public Builder(Tracer tracer, String spanName) {
+      builder = tracer.spanBuilder(spanName);
+    }
+
+    public SpanEx.Builder setParent(SpanEx parent) {
+      builder.setParent(parent.span);
+      return this;
+    }
+
+    public SpanEx.Builder setParent(SpanContext remoteParent) {
+      builder.setParent(remoteParent);
+      return this;
+    }
+
+    public SpanEx.Builder setNoParent() {
+      builder.setNoParent();
+      return this;
+    }
+
+    // option 1... an interface that encapsulates the attribute nature
+    public SpanEx.Builder addLink(SpanContext context, Attributes attributes) {
+      builder.addLink(context, makeAttributeMap(attributes));
+      return this;
+    }
+
+    // option 2... a type-safe key-value pair with varargs parameters
+    public SpanEx.Builder addLink(SpanContext context, Attribute... attributes) {
+      builder.addLink(context, makeAttributeMap(attributes));
+      return this;
+    }
+
+    // option 3... provide a concrete Link class.
+    public SpanEx.Builder addLink(LinkEx link) {
+      return addLink(link.getSpanContext(), link.getAttributes());
+    }
+
+    public SpanEx.Builder setSpanKind(Kind spanKind) {
+      builder.setSpanKind(spanKind);
+      return this;
+    }
+
+    public SpanEx.Builder setStartTimestamp(long startTimestamp) {
+      builder.setStartTimestamp(startTimestamp);
+      return this;
+    }
+
+    public SpanEx startSpan() {
+      return new SpanEx(builder.startSpan());
+    }
+
+    // option 1... provide a single attribute
+    public SpanEx.Builder setAttribute(Attribute attribute) {
+      builder.setAttribute(attribute.key().key(), makeValue(attribute));
+      return this;
+    }
+
+    // option 2... a type-safe key-value pair with varargs parameters
+
+    /** doc me. */
+    public SpanEx.Builder setAttribute(Attribute... attributes) {
+      for (Attribute attribute : attributes) {
+        builder.setAttribute(attribute.key().key(), makeValue(attribute));
+      }
+      return this;
+    }
+
+    // option 3... accept a bunch of attributes at once
+
+    /** doc me. */
+    public SpanEx.Builder setAttribute(Attributes attributes) {
+      for (AttributeKey key : attributes.getKeys()) {
+        builder.setAttribute(key.key(), makeValue(attributes, key));
+      }
+      return this;
+    }
```

Comment:
comparing

```
spanBuilder.setAttributes(
        MapBasedAttributes.newBuilder().put(key1, value1)
            .put(key2, value2)
            .put(key3, value3)
            .build())
```

and

```
spanBuilder.setAttribute(key1, value1)
           .setAttribute(key2, value2)
           .setAttribute(key3, value3);
```

the latter is more efficient, and seems more user-friendly to me

Reply:
> i think i prefer the more efficient and more user-friendly `setAttribute(key, value)`.
> 
> from a user perspective, i don't think the overloaded methods make the API surface feel bigger (they only add to the API surface from a maintainer's perspective).

I agree with you that we should offer the `setAttribute(key, value)` but I think we should also offer the 3rd option using `Attributes`. In many semantic conventions, say HTTP, many attributes don't change. For example, if a library has to manually instrument its code to serve an HTTP request, the attributes `method`, `url`, `target`, ... do not change between requests so they can be cached.

Reply:
This is an interesting thought, yes. Pre-bundling the attributes that are relevant for a given HTTP endpoint is possible, but is that likely to be a common use-case, as opposed to more generic instrumentation that would need to dynamically generate the values?

Reply:
Would caching an `Attributes` object give some performance benefit? Or is it just for convenience? Maybe a concrete example would help to understand how caching the `Attributes` would be more convenient compared to settings the attributes directly.

Reply:
I was thinking out loud without having a concrete example in mind. Thinking more thoroughly on a possible usage, I would use the `Attributes` interface to decouple span creation and attribute definition, similarly to how it's used here: 

https://github.com/newrelic-forks/opentelemetry-java/blob/37c87b907289de35be6dfe4c3aee132e1815b7db/contrib/api_incubator/src/test/java/io/opentelemetry/trace/TracerExTest.java#L84-L85

Reply:
The decoupling is nice in that example, but if the attribute values aren't static, I think you'd need to pass in 8 parameters to that `httpAttribute()` method, at which point it may not be so nice.